// @tronweb3/tronwallet-abstract-adapter/lib/esm/errors.js
export class WalletError extends Error {
    constructor(message, error) {
        super(message);
        this.error = error;
    }
}
/**
 * Occurs when wallet is not installed.
 */
export class WalletNotFoundError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletNotFoundError';
        this.message = 'The wallet is not found.';
    }
}
/**
 * Occurs when connect to a wallet but there is no wallet selected.
 */
export class WalletNotSelectedError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletNotSelectedError';
        this.message = 'No wallet is selected. Please select a wallet.';
    }
}
/**
 * Occurs when wallet is disconnected.
 * Used by some wallets which won't connect automatically when call `signMessage()` or `signTransaction()`.
 */
export class WalletDisconnectedError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletDisconnectedError';
        this.message = 'The wallet is disconnected. Please connect first.';
    }
}
/**
 * Occurs when try to connect a wallet.
 */
export class WalletConnectionError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletConnectionError';
    }
}
/**
 * Occurs when try to disconnect a wallet.
 */
export class WalletDisconnectionError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletDisconnectionError';
    }
}
/**
 * Occurs when call `signMessage()`.
 */
export class WalletSignMessageError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletSignMessageError';
    }
}
/**
 * Occurs when call `signTransaction()`.
 */
export class WalletSignTransactionError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletSignTransactionError';
    }
}
/**
 * Occurs when load wallet
 */
export class WalletWalletLoadError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletWalletLoadError';
    }
}
/**
 * Occurs when walletconnect QR window is closed.
 */
export class WalletWindowClosedError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletWindowClosedError';
        this.message = 'The QR window is closed.';
    }
}
/**
 * Occurs when request wallet to switch chain.
 */
export class WalletSwitchChainError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletSwitchChainError';
    }
}
/**
 * Occurs when get network infomation.
 */
export class WalletGetNetworkError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletGetNetworkError';
    }
}
