// @tronweb3/tronwallet-abstract-adapter/lib/cjs/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletGetNetworkError = exports.WalletSwitchChainError = exports.WalletWindowClosedError = exports.WalletWalletLoadError = exports.WalletSignTransactionError = exports.WalletSignMessageError = exports.WalletDisconnectionError = exports.WalletConnectionError = exports.WalletDisconnectedError = exports.WalletNotSelectedError = exports.WalletNotFoundError = exports.WalletError = void 0;
class WalletError extends Error {
    constructor(message, error) {
        super(message);
        this.error = error;
    }
}
exports.WalletError = WalletError;
/**
 * Occurs when wallet is not installed.
 */
class WalletNotFoundError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletNotFoundError';
        this.message = 'The wallet is not found.';
    }
}
exports.WalletNotFoundError = WalletNotFoundError;
/**
 * Occurs when connect to a wallet but there is no wallet selected.
 */
class WalletNotSelectedError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletNotSelectedError';
        this.message = 'No wallet is selected. Please select a wallet.';
    }
}
exports.WalletNotSelectedError = WalletNotSelectedError;
/**
 * Occurs when wallet is disconnected.
 * Used by some wallets which won't connect automatically when call `signMessage()` or `signTransaction()`.
 */
class WalletDisconnectedError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletDisconnectedError';
        this.message = 'The wallet is disconnected. Please connect first.';
    }
}
exports.WalletDisconnectedError = WalletDisconnectedError;
/**
 * Occurs when try to connect a wallet.
 */
class WalletConnectionError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletConnectionError';
    }
}
exports.WalletConnectionError = WalletConnectionError;
/**
 * Occurs when try to disconnect a wallet.
 */
class WalletDisconnectionError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletDisconnectionError';
    }
}
exports.WalletDisconnectionError = WalletDisconnectionError;
/**
 * Occurs when call `signMessage()`.
 */
class WalletSignMessageError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletSignMessageError';
    }
}
exports.WalletSignMessageError = WalletSignMessageError;
/**
 * Occurs when call `signTransaction()`.
 */
class WalletSignTransactionError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletSignTransactionError';
    }
}
exports.WalletSignTransactionError = WalletSignTransactionError;
/**
 * Occurs when load wallet
 */
class WalletWalletLoadError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletWalletLoadError';
    }
}
exports.WalletWalletLoadError = WalletWalletLoadError;
/**
 * Occurs when walletconnect QR window is closed.
 */
class WalletWindowClosedError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletWindowClosedError';
        this.message = 'The QR window is closed.';
    }
}
exports.WalletWindowClosedError = WalletWindowClosedError;
/**
 * Occurs when request wallet to switch chain.
 */
class WalletSwitchChainError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletSwitchChainError';
    }
}
exports.WalletSwitchChainError = WalletSwitchChainError;
/**
 * Occurs when get network infomation.
 */
class WalletGetNetworkError extends WalletError {
    constructor() {
        super(...arguments);
        this.name = 'WalletGetNetworkError';
    }
}
exports.WalletGetNetworkError = WalletGetNetworkError;
